src/operator/ndarray_op.cc

```diff
@@ -80,40 +83,43 @@ void NDArrayOp<xpu>::Backward(const OpContext &ctx,
                     const std::vector<TBlob> &aux_args) {
   using namespace mshadow;
   Context ndctx = get_ctx();
-  std::vector<void*> ptrs;
+  std::vector<void*> *ptrs = new std::vector<void*>;
   std::vector<Engine::VarHandle> ndvar;
-  std::vector<int> tags;
+  std::vector<int> *tags = new std::vector<int>;
   for (auto& i : req) CHECK_NE(i, kAddTo);
 
   for (auto& blob : in_data) {
-    ptrs.push_back(reinterpret_cast<void*>(new NDArray(blob, ndctx.dev_id)));
-    tags.push_back(0);
+    ptrs->push_back(reinterpret_cast<void*>(new NDArray(blob, ndctx.dev_id)));
+    tags->push_back(0);
   }
   for (auto& blob : out_data) {
-    ptrs.push_back(reinterpret_cast<void*>(new NDArray(blob, ndctx.dev_id)));
-    tags.push_back(1);
+    ptrs->push_back(reinterpret_cast<void*>(new NDArray(blob, ndctx.dev_id)));
+    tags->push_back(1);
   }
   for (auto& blob : in_grad) {
     NDArray* nd = new NDArray(blob, ndctx.dev_id);
-    ptrs.push_back(reinterpret_cast<void*>(nd));
+    ptrs->push_back(reinterpret_cast<void*>(nd));
     ndvar.push_back(nd->var());
-    tags.push_back(2);
+    tags->push_back(2);
   }
   std::sort(ndvar.begin(), ndvar.end());
   ndvar.resize(std::unique(ndvar.begin(), ndvar.end()) - ndvar.begin());
   for (auto& blob : out_grad) {
-    ptrs.push_back(reinterpret_cast<void*>(new NDArray(blob, ndctx.dev_id)));
-    tags.push_back(3);
+    ptrs->push_back(reinterpret_cast<void*>(new NDArray(blob, ndctx.dev_id)));
+    tags->push_back(3);
   }
 
   std::vector<NDArray> ndcpy;
-  for (auto& i : ptrs) {
+  for (auto& i : *ptrs) {
     ndcpy.push_back(*reinterpret_cast<NDArray*>(i));
   }
 
-  CHECK(param_.pinfo->backward(ptrs.size(), ptrs.data(), tags.data(), param_.pinfo->p_backward));
-  Engine::Get()->PushSync([ndcpy, ctx](RunContext rctx){ ctx.async_on_complete(); },
-                          ndctx, ndvar, {});
+  CHECK(param_.pinfo->backward(ptrs->size(), ptrs->data(), tags->data(), param_.pinfo->p_backward));
+  Engine::Get()->PushSync([ptrs, tags, ndcpy, ctx](RunContext rctx){
+      ctx.async_on_complete();
```

Comment:
when you call PushSync, there is no need to call async_on_compelte, this is only reserved for the async call in the operators